File: App/Resources/AlgorithmDetails/patiencesort/patiencesort.rb
class MinHeap
  def initialize
    @storage = []
  end

  def empty?
    @storage.empty?
  end

  def push(entry)
    @storage.push(entry)
    i = @storage.length - 1
    while i > 0
      parent = (i - 1) / 2
      break if @storage[parent][0] <= @storage[i][0]

      @storage[parent], @storage[i] = @storage[i], @storage[parent]
      i = parent
    end
  end

  def pop_min
    result = @storage[0]
    last = @storage.pop
    unless @storage.empty?
      @storage[0] = last
      i = 0
      loop do
        left = (2 * i) + 1
        right = (2 * i) + 2
        smallest = i
        smallest = left if left < @storage.length && @storage[left][0] < @storage[smallest][0]
        smallest = right if right < @storage.length && @storage[right][0] < @storage[smallest][0]
        break if smallest == i

        @storage[i], @storage[smallest] = @storage[smallest], @storage[i]
        i = smallest
      end
    end
    result
  end
end

def sort(arr)
  n = arr.length
  piles = []
  tops = []

  arr.each do |x|
    # binary search: leftmost pile whose top is >= x
    lo = 0
    hi = piles.length
    while lo < hi
      mid = (lo + hi) / 2
      if tops[mid] >= x
        hi = mid
      else
        lo = mid + 1
      end
    end
    if lo == piles.length
      piles.push([x])
      tops.push(x)
    else
      piles[lo].push(x)
      tops[lo] = x
    end
  end

  heap = MinHeap.new
  piles.each_index do |i|
    heap.push([tops[i], i])
  end

  result = []
  until heap.empty?
    _top_value, pile_index = heap.pop_min
    value = piles[pile_index].pop
    result.push(value)
    heap.push([piles[pile_index].last, pile_index]) unless piles[pile_index].empty?
  end

  (0...n).each do |i|
    arr[i] = result[i]
  end
end

array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
p array

File: App/Resources/AlgorithmDetails/patiencesort/patiencesort.js
class MinHeap {
  constructor() {
    this.storage = [];
  }

  get isEmpty() {
    return this.storage.length === 0;
  }

  push(entry) {
    this.storage.push(entry);
    let i = this.storage.length - 1;
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (this.storage[parent].top <= this.storage[i].top) break;
      [this.storage[parent], this.storage[i]] = [
        this.storage[i],
        this.storage[parent],
      ];
      i = parent;
    }
  }

  popMin() {
    const result = this.storage[0];
    const last = this.storage.pop();
    if (this.storage.length > 0) {
      this.storage[0] = last;
      let i = 0;
      while (true) {
        const left = 2 * i + 1;
        const right = 2 * i + 2;
        let smallest = i;
        if (
          left < this.storage.length &&
          this.storage[left].top < this.storage[smallest].top
        )
          smallest = left;
        if (
          right < this.storage.length &&
          this.storage[right].top < this.storage[smallest].top
        )
          smallest = right;
        if (smallest === i) break;
        [this.storage[i], this.storage[smallest]] = [
          this.storage[smallest],
          this.storage[i],
        ];
        i = smallest;
      }
    }
    return result;
  }
}

function sort(arr) {
  const n = arr.length;
  const piles = [];
  const tops = [];

  for (const x of arr) {
    // binary search: leftmost pile whose top is >= x
    let lo = 0;
    let hi = piles.length;
    while (lo < hi) {
      const mid = (lo + hi) >> 1;
      if (tops[mid] >= x) {
        hi = mid;
      } else {
        lo = mid + 1;
      }
    }
    if (lo === piles.length) {
      piles.push([x]);
      tops.push(x);
    } else {
      piles[lo].push(x);
      tops[lo] = x;
    }
  }

  const heap = new MinHeap();
  for (let i = 0; i < piles.length; i++) {
    heap.push({ top: tops[i], pileIndex: i });
  }

  const result = [];
  while (!heap.isEmpty) {
    const entry = heap.popMin();
    const value = piles[entry.pileIndex].pop();
    result.push(value);
    if (piles[entry.pileIndex].length > 0) {
      heap.push({
        top: piles[entry.pileIndex][piles[entry.pileIndex].length - 1],
        pileIndex: entry.pileIndex,
      });
    }
  }

  for (let i = 0; i < n; i++) {
    arr[i] = result[i];
  }
}

var array = [0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56];
sort(array);
console.log("[" + array.join(", ") + "]");
